### ISSUE TITLE
Easier, More Regular Releases 

### ISSUE THREAD

**Issue by saulshanabrook** [jupyterlab/jupyterlab#8195]
It would be useful to be able to do releases more regularly and to be able to do them with less effort. To do the last minor and patches release, @jasongrout and I estimated that it took each of us between 15 and 30 hours.

Ideally, we should drive this cost down to a minimal amount and be able to spread it over more people and more time so that we are able to make releases more regularly.

The time to release was split between a number of tasks:

1. Determining if each PR is a minor or patch change.
2. Make sure there are proper milestones and branches for each new release.
3. Make sure all PRs and issues are assigned to the correct milestone
4. Backport any PRs that we need to other branches.
5. doing release, which means:
   1. Creating clean environment
   2. Do JS releases and build python release
   3. Test release by opening jlab
   4. Push Python release, push git tags
8. Submit PRs to conda to trigger a release.
9. Build changelog based on PRs on the milestone

The next steps here is to try to come up with a plan that can move the work for all these tasks around or automated them, so doing release, ideally, involves maybe triggering a github action and merging some PRs.

**Comment by jasongrout** [comment]
We could build a workflow script that automatically assigns PRs (and relevant issues?) to the appropriate milestone based on a mapping between milestones and commit ranges.

**Comment by saulshanabrook** [comment]
> We could build a workflow script that automatically assigns PRs (and relevant issues?) to the appropriate milestone based on a mapping between milestones and commit ranges.

Yes and also based on major/minor/patch tags on the issue, which should give the option of backports to open.

**Comment by blink1073** [comment]
We could use api extractor to automatically determine if a PR has any backward compatibility changes and assign it a label.

**Comment by blink1073** [comment]
For patch changes, on merge, it could automatically invoke meeseeksdev to backport to the previous milestone (stored in config).

**Comment by blink1073** [comment]
If we had config for branches that are designated as major, minor, and patch targets, then merging a PR could target all of them.

**Comment by blink1073** [comment]
Including applying the milestone automatically

**Comment by saulshanabrook** [comment]
> If we had config for branches that are designated as major, minor, and patch targets, then merging a PR could target all of them.

We could also make master always be the next major release, so everything is merged into master, and then if its not a major change, we created a backport PR to the other branches that make sense. i.e. if its patch maybe the most recent patch release plus we choose to backport it to.

**Comment by blink1073** [comment]
That sounds right.  What I mean is that we encode that decision in a config file.

**Comment by jasongrout** [comment]
+1. Right now the milestone commit ranges are encoded at https://github.com/jupyterlab/jupyterlab/blob/44f161b1025acb1948599effdf55b28dcd9693c5/scripts/milestone_check.py#L13-L26

**Comment by saulshanabrook** [comment]
> That sounds right. What I mean is that we encode that decision in a config file.

Yeah or we make the branches/milestone regular enough so we can infer it. It might be nice not to have a file that we have to keep up to date.

**Comment by saulshanabrook** [comment]
I opened up a PR against team-compass to articulate a possible plan here: https://github.com/jupyterlab/team-compass/pull/54

Would love any feedback.
